egs/wsj/s5/steps/dict/train_g2p.sh

```diff
@@ -45,7 +45,7 @@ mkdir -p $wdir/log
 # Optionally remove words that are mapped to a single silence phone from the lexicon.
 if $only_words && [ ! -z "$silence_phones" ]; then
   awk 'NR==FNR{a[$1] = 1; next} {s=$2;for(i=3;i<=NF;i++) s=s" "$i;a[$1]=s;if(!(s in a)) print $1" "s}' \
-    $silence_phones > $wdir/lexicon_onlywords.txt
+    $lexicon $silence_phones > $wdir/lexicon_onlywords.txt
```

Comment:
I know this awk script was pre-existing, but it's rather ugly.   Can you rewrite it to make it clearer?  I prefer perl or awk to avoid a super-long python script.  Also the script seems to remove words where the pronunciation was seen before, which I doubt was the intention.  And it's not clear to me that the order of 'lexicon' vs. 'silence_phones' as you have it, is correct anyway.   Bear in mind that it's possible that there could be multiple silence phones on a single line, in the $silence_phones file, which I think this script does not handle properly.
You can use the following as inspiration (reading from a file directly is less ugly than the FNR trick):
```
# Get training data with OOV words (w.r.t. our current vocab) replaced with <unk>,
# as well as adding </s> symbols at the end of each sentence
cat $train_text | awk -v w=$dir/wordlist.all \
  'BEGIN{while((getline<w)>0) v[$1]=1;}
  {for (i=2;i<=NF;i++) if ($i in v) printf $i" ";else printf "<unk> ";print ""}' | sed 's=$= </s>=g' \
  | utils/shuffle_list.pl | gzip -c > $dir/all.gz
```

Reply:
@xiaohui-zhang would have idea about this part.

Reply:
@danpovey do we allow multiple silence phones to occur in the same row in data/local/dict/silence_phones.txt ? I thought they have to be put in different columns.